### PR TITLE
net: tcp: Check TCP option count before exit

### PR DESCRIPTION
**Pull request by ghost**
While parsing TCP options, all options field should be read/processed and
the TCP option count should be zero before exit.
But if TCP option count is non-zero, which means there is unread options, and
the option field can be malformed, and the function should return with error.

Signed-off-by: Tedd Ho-Jeong An <tedd.an@intel.com>

### CODE REVIEW

**Review by pfalcon — CHANGES_REQUESTED** [review]
Note that before `if (opt_totlen)` that you add, there's `while (opt_totlen)`. It has a single break - on NET_TCP_END_OPT. So, this effectively tests that there're no more octets beyond NET_TCP_END_OPT in options. Fairly speaking, I'd take time to re-read the TCP standard to find a place where it explicitly prohibits that. Yes, there're other ways to pad things, but then NET_TCP_END_OPT itself is also superfluous. The only reason to have NET_TCP_END_OPT is exactly to force stop of the option processing. Consider a router/other appliance which takes care of option processing itself. To avoid heavy packet rewriting (memory movements), it just stuffs NET_TCP_END_OPT there, and leave everything else in place. So, let me be -1 on this. At most, this could be a warning.

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -1406,6 +1406,15 @@ int net_tcp_parse_opts(struct net_pkt *pkt, int opt_totlen,
 		opt_totlen -= optlen;
 	}
 
+	/* All options must be read/processed and option count should be zero.
+	 * If the option count is non-zero, which means there is unread options
```
Comment (pfalcon) [inline subsys/net/ip/tcp.c]: Your comment (and commit message btw too) talks about "option count" (apparently counted in options), but actually deals with option length (apparently counted in bytes). It's not good when comments don't match the code.

**Comment by bluespring** [comment]
@pfalcon 

Thanks for your comment.
I got your point and realized that my commit couldn't handle the case if there was any extra bytes in the padding after NET_TCP_END_OPT.

My focus was when the options are malformed, especially when it is not following the spec or corrupted in the middle. 
For example, if the options are: 
`1B 07 00 00 00 00 00 00   02 04 05 A0 04 02 08 0A`
`6E 98 04 55 00 00 00 00 `
`1B (Quick Start Response)` requires 8 bytes of data but if the length filed is `07` for some reason, our code reads the 8th `00` as NET_TCP_END_OPT and it will returns success and reply ACK even though the options are malformed.
Although I am not sure what's the possible consequence of sending ACK after these kinds of malformed options in SYN packet, I think it shouldn't reply ACK basically.

So, what do you think about these kinds of "malformed" packet and how do we suppose to handle it?
I understand that we don't check the format of all known options except MSS, but do you think it is worth to try to check the format of all known options even if we are not really using it?

P.S. I will be more careful on commit message and comment in next time. Thanks for point out that also. :)

**Comment by jukkar** [comment]
> So, what do you think about these kinds of "malformed" packet and how do we suppose to handle it?
> I understand that we don't check the format of all known options except MSS, but do you think it is worth to try to check the format of all known options even if we are not really using it?

I suppose this could be handled in similar way as how IPv6 is handling options. If the option is a known one to us and we find error in it, then the packet is dropped. If the option is unknown, we obviously cannot detect any errors in it and skip it.

**Comment by pfalcon** [comment]
I agree with @jukkar here - if we handle an option, i.e. it's useful to Zephyr's IP stack, then we'll be able to catch violations like an example you give with 1B (Quick Start Response), but with length 07. But if not, we can just following generic option format of (code, length).

Also, what exactly we're trying protect against? If it's random packet corruption during transfer, it would be caught by packet checksums. And if it's  someone deliberately sends wrong option lengths, as long as they can't use that to mount attack against us (i.e. we don't use those length blindly, but verify it), it shouldn't be a problem either.

**Pull request closed without merging**